zerver/tornado/event_queue.py

```diff
@@ -970,7 +972,9 @@ def get_client_payload(apply_markdown: bool, client_gravatar: bool) -> Dict[str,
         # shouldn't receive notifications even if they were online. In that case we can
         # avoid the more expensive `receiver_is_off_zulip` call, and move on to process
         # the next user.
-        if not user_notifications_data.is_notifiable(acting_user_id=sender_id, idle=True):
+        if is_sender_welcome_bot or not user_notifications_data.is_notifiable(
+            acting_user_id=sender_id, idle=True
+        ):
```

Comment:
We likely need a similar check in `missedmessage_hook`... or perhaps it'd be better to make `maybe_enqueue_notifications` accept this parameter and do this check.

Reply:
As per [docs](https://zulip.readthedocs.io/en/latest/subsystems/notifications.html#the-mobile-email-notifications-flow):
> We try to contain the “when to notify” business logic in the zerver/lib/notification_data.py class methods.

So, ideally the logic should reside in the [`is_email_notifiable()`](https://github.com/zulip/zulip/blob/cebf43903c5fc676432a600281fb18a7e192f3af/zerver/lib/notification_data.py#L136) function.

Possible solutions:

1. In `maybe_enqueue_notification`, update [this logic ](https://github.com/zulip/zulip/blob/cebf43903c5fc676432a600281fb18a7e192f3af/zerver/tornado/event_queue.py#L830)as:

> if **not is_sender_welcome_bot and** user_notifications_data.is_email_notifiable(acting_user_id, idle):

2. pass `is_sender_welcome_bot` as the third argument to `is_email_notifiable` and perform the check inside the function. (update test cases accordingly)

3. fetch `welcome_bot` id inside the `is_email_notifiable` and perform the check. But this will lead to an added query. So, this should be avoided.

I think 2nd approach is the best. Do you see any shortcomings?

Reply:
@abhijeetbodas2001 @PIG208 @akashaviator I'd be curious for your thoughts on where this should sit.

Reply:
IMO, `is_email_notifiable()` should check this. It should always be able to return the result it is supposed to give.

Reply:
Yeah I think I agree with that reasoning. @prakhar1144 can you try reworking this to that model?

Reply:
I think we can extract a separate method `should_notify` or something, from generic checks that both `is_push_notifiable` and `is_email_notifiable` [do](https://github.com/zulip/zulip/blob/main/zerver/lib/notification_data.py#L113-L120)), which includes the welcome bot check we are planning for right now.

Reply:
After considering everyone's suggestions. This is what I'm planning:

* A separate method named `should_notify` or `should_sender_notify` extracts the generic checks i.e 
```
  if self.user_id == acting_user_id:
      return None

  if self.sender_is_muted:
      return None

  if is_sender_welcome_bot:
      return None
```
* Pass `is_sender_welcome_bot` as the last argument to `is_notifiable` and `maybe_enqueue_notifications`.

* In the case of `process_message_event`, the check will stop [here itself](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L973) if `sender_is_welcome_bot` thus avoiding the [expensive operation](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L969)

* In the case of [`missedmessage_hook`](https://github.com/zulip/zulip/blob/8883b643881f0f243712ad020d22ebd400fb0b94/zerver/tornado/event_queue.py#L716), `maybe_enqueue_notifications` will take care of the checks.

Let me know if it sounds good to you guys.

Reply:
@timabbott @PIG208 

Please leave feedback so I can proceed. It would be greatly appreciated.

Reply:
> I think we can extract generic checks that both `is_push_notifiable` and `is_email_notifiable` [have](https://github.com/zulip/zulip/blob/main/zerver/lib/notification_data.py#L116-L120), and create a separate method called `should_notify` or something. This helper should then include the welcome bot check we are planning for right now.

That makes sense to me, except that we could use something more descriptive than `should_notify` (like `trivially_should_not_notify` or something like that, since it will contain trivial checks like muted user/welcome bot which don't involve any notification settings).
@prakhar1144 you can probably add a prep commit to pull out the already existing common checks into a separate function and then add the welcome bot check to that function.

Reply:
One question I have is -- why not make `is_sender_welcome_bot` a property of `UserMessageNotificationsData` itself? For context, one of the motivations behind making `UserMessageNotificationsData` was that having so many parameters for notification checks had become unwieldy in `event_queue.py` :)

Reply:
This sounds right to me.

Reply:
> One question I have is -- why not make `is_sender_welcome_bot` a property of `UserMessageNotificationsData` itself? For context, one of the motivations behind introducing `UserMessageNotificationsData` was that having so many parameters for notification checks and passing them around functions in `event_queue.py` was becoming unwieldy :)

Hm, Makes Sense!